sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -585,25 +585,26 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
 
   if (NumDevices)
     *NumDevices = ZeDeviceCount;
-
-  try {
-    // TODO: Delete array at teardown
-    ze_device_handle_t *ZeDevices = new ze_device_handle_t[ZeDeviceCount];
-    ZE_CALL(zeDeviceGet(ZeDriver, &ZeDeviceCount, ZeDevices));
-
-    for (uint32_t I = 0; I < ZeDeviceCount; ++I) {
-      if (I < NumEntries) {
-        Devices[I] = new _pi_device(ZeDevices[I], Platform);
-        pi_result Result = Devices[I]->initialize();
-        if (Result != PI_SUCCESS) {
-          return Result;
+  if (NumEntries > 0) {
+    try {
+      // TODO: Delete array at teardown
```

Comment:
We still have a leak here since the memory allocated here is never deleted. This should probably wait until proper tear-down is implemented.

Even more importantly every call to `cl::sycl::device::get_devices()` allocates new memory. We should probably cache discovered PI platforms/devices so that subsequent calls to `get_platforms()` and `get_devices()` do not consume more memory.


Tagging @glyons-intel who was looking into similar issue I believe